---
title: "16S Preparatory Analysis"
output: html_document
date: "2024-06-11"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/DADS")

library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(readr); packageVersion("readr")
library(purrr); packageVersion("purrr")
library(stringr); packageVersion("stringr")
library(dplyr); packageVersion("dplyr")
library(igraph); packageVersion("igraph")
library(metacoder); packageVersion("metacoder")
library(decontam); packageVersion("decontam")
library(data.table); packageVersion("data.table")
library(dada2); packageVersion("dada2")
```

### Load data
```{r load files, include=TRUE}
#Load DADA2 files
seqtab_nochim <- read.table("Data/16S/dada2_outputs/16S_seqtab_nochim.out",sep="\t",stringsAsFactors=F,header=T)

taxa <- read.table("Data/16S/dada2_outputs/16S_taxa_species_separate_minboot60.out",sep="\t",stringsAsFactors=F,header=T)

revise_species <- function(genus, species) {
  # Combine genus and species with an underscore if species is not NA
  revised_species <- ifelse(is.na(species), NA, paste(genus, species, sep = "_"))
  return(revised_species)
}

# Apply the function to revise Species column
taxa$Species <- revise_species(taxa$Genus, taxa$Species)

samdf <- read.csv("Data/16S/metadata_dbs/Metadata_16S_all.csv")
```

### Track number of reads throughout the pipeline
```{r track reads through DADA2 analysis, include=TRUE}
track <- read.table("Data/16S/dada2_outputs/16S_trackreads.out", sep="\t", stringsAsFactors = F, header=T)

track_trim<-track[!(track$Sample %in% c("emptyextraction", "HD4390trim-S1", "HD4397trim-S15", "NTC")), ]
```

### Format dataframes to create phyloseq object
```{r format dataframes, include=TRUE}
# # For ease of analysis, convert the DADA2 outputs into a Phyloseq object, and filter out any necessary samples
all(rownames(seqtab_nochim) %in% samdf$Sample_name16S) 
rownames(samdf) <- samdf$Sample_name16S

samdata.ps <- sample_data(samdf)

taxa.ps <- tax_table(as.matrix(taxa))

seqtab.nochim.ps <- otu_table(seqtab_nochim, taxa_are_rows=FALSE)
ps <- phyloseq(seqtab.nochim.ps, samdata.ps, taxa.ps)

#Add refseq slot
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

#I randomly subsampled from two samples that had way more reads. For now I will remove these samples, but it is always an option to use the subsampled samples
ps_filtered<-subset_samples(ps, ps@sam_data@row.names != "HD4397trim-S15")
ps_filtered<-subset_samples(ps_filtered, ps@sam_data@row.names != "HD4390trim-S1")
```

### Decontamination step using decontam package-this was most necessary for 16S analysis.
```{r decontam, include=TRUE}
# Inspect library sizes
df <- as.data.frame(sample_data(ps_filtered)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps_filtered)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=Sample_or_control)) + geom_point()

sample_data(ps_filtered)$is.neg <- sample_data(ps_filtered)$Sample_or_control == "control"
contamdf.prev <- isContaminant(ps_filtered, method="prevalence", neg="is.neg")
table(contamdf.prev$contaminant)
head(which(contamdf.prev$contaminant))

contamdf.prev05 <- isContaminant(ps_filtered, method="prevalence", neg="is.neg", threshold=0.5)
table(contamdf.prev05$contaminant)
which(contamdf.prev05$contaminant)

ps.pa <- transform_sample_counts(ps_filtered, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$Sample_or_control == "control", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Sample_or_control == "sample", ps.pa)

# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prev05$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

#Remove contaminants
ps.decontam <- prune_taxa(!contamdf.prev05$contaminant, ps_filtered)

ps.decontam <- ps.decontam %>%
  subset_taxa(
    Kingdom == "Bacteria" &
      Family  != "Mitochondria" &
      Class   != "Chloroplast"
  )

save(ps.decontam, file = "Data/16S/filtered_data/16S_phyloseq.decontam.RData")
```


### Take a first glance at the data

Helpful for deciding how to filter data. Prevelance is how many samples have the ASV, and total abundance is how many times ASV is present across all samples. I don't do any filtering until I split up samples into relevant experiment.
```{r examine ASV Prevalance, include=TRUE}
#Prevalence-number of samples in which a taxon appears at least once
#Total abundance
prevdf = apply(X = phyloseq::otu_table(ps.decontam),
               MARGIN = ifelse(taxa_are_rows(ps.decontam), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps.decontam),
                    tax_table(ps.decontam))
```


### Subset data so we just have access to DADS samples

Now I will focus on DADS experiment and prepare dataset for further analysis by subsetting these samples from the rest of the sample data. I will also remove allow abundance ASVs not present in at least one sample and not present at least five times
```{r Subset dataset, include=TRUE}
dads.ps<-subset_samples(ps.decontam, Experiment=="dunglabsamples")
dads.ps<- prune_taxa(taxa_sums(dads.ps) > 0, dads.ps)

readsums<-sort(sample_sums(dads.ps))
median(readsums)

dads.ps@sam_data$Treatment <- factor(dads.ps@sam_data$Treatment,
                                     levels = c("untreated", "DADS_1gal_A"),
                                     labels = c("untreated", "DADS_treated"))

dads.ps@sam_data$Incubation <- factor(dads.ps@sam_data$Incubation,
                                      levels = c("aerobic", "anaerobic"))

dads.ps@sam_data$Soil <- factor(dads.ps@sam_data$Soil,
                                     levels = c("madras", "tulelake"))

#Let's filter an ASVs not present at least 5 times
dads_5asv.ps <- prune_taxa(taxa_sums(dads.ps) > 5, dads.ps)
dads_5asv.ps<- prune_taxa(taxa_sums(dads_5asv.ps) > 0, dads_5asv.ps)

save(dads.ps, file="Data/16S/filtered_data/dadsexp.RData")
save(dads_5asv.ps, file="Data/16S/filtered_data/dads_5asv.RData")

readsums<-sort(sample_sums(dads.ps))
median(readsums)
summary(readsums)
```


### Make prevalence tables to better examine low abundance ASVs
```{r check prevalence dataframes after filterig}
#Double check that the filtering makes sense by making new prevalence dataframe
prevdf_check = apply(X = phyloseq::otu_table(dads_5asv.ps),
               MARGIN = ifelse(taxa_are_rows(dads_5asv.ps), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf_check = data.frame(Prevalence = prevdf_check,
                    TotalAbundance = taxa_sums(dads_5asv.ps),
                    tax_table(dads_5asv.ps))

```

### Collect read count info for just the DADS samples
```{r subset tracked reads}
samples_to_subset <- samdf %>%
  filter(Experiment=="dunglabsamples") %>% 
  pull(Sample_name16S)
# Filter the track dataframe

#track$sample <- rownames(track)
subset_track_reads <- track[rownames(track) %in% samples_to_subset, ]
write.csv(subset_track_reads, file="Data/16S/filtered_data/16S_subset_track_reads_dads.csv",row.names = TRUE)
# Print summary of the filtered track dataframe
summary(subset_track_reads)
```

### Create prevalance plots (Rel abundance vs. total abundance) facetted by phylum
```{r prevalence plots organized by phylum}
prevdf1 = subset(prevdf_check, Phylum %in% get_taxa_unique(dads_5asv.ps, "Phylum"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(dads_5asv.ps),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.005, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
```

### Make histograms
```{r Reads sequenced by experiment}
#Before filtering reads
sdt = data.table::data.table(as(sample_data(dads.ps), "data.frame"),
                 TotalReads = sample_sums(dads.ps), keep.rownames = TRUE)
pseqDepth = ggplot(sdt, aes(TotalReads)) + geom_histogram(color="black", fill="lightblue") + ggtitle("Histogram of sequencing depth-before filtering samples with low read depths") + labs(x="Depth", y="Frequency")
pseqDepth

#After filtering reads
sdt_filt = data.table::data.table(as(sample_data(dads_5asv.ps), "data.frame"),
                 TotalReads = sample_sums(dads_5asv.ps), keep.rownames = TRUE)
pseqDepth_filt = ggplot(sdt_filt, aes(TotalReads)) + geom_histogram(color="black", fill="lightblue") + ggtitle("Histogram of sequencing depth after filtering samples with low read depths") + labs(x="Depth", y="Frequency")
pseqDepth_filt
```

### Software used
```{r, software used}
sessionInfo()
```
