title: "DADS_compiled_figures"
output: html_document
date: "2024-06-11"
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/SOD_DADS_microbiome")

library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(readr); packageVersion("readr")
library(purrr); packageVersion("purrr")
library(furrr); packageVersion("furrr")
library(dplyr); packageVersion("dplyr")
library(stringr); packageVersion("stringr")
library(metacoder); packageVersion("metacoder")
library(data.table); packageVersion("data.table")
library(Biostrings); packageVersion("Biostrings")
library(vegan);packageVersion("vegan")
library(agricolae);packageVersion("agricolae")
library(magrittr);packageVersion("magrittr")
library(data.table);packageVersion("data.table")
library(DESeq2);packageVersion("DESeq2")
library(cowplot);packageVersion("cowplot")
library(doParallel);packageVersion("doParallel")
```

### Load data from all three experiments
```{r subset data ITS, include=TRUE}
load("Data/ITS/filtered_data/dadsexp.RData")
dads_its.ps<-dads.ps

load("Data/ITS/filtered_data/dads_5asv.RData")
dads_5asv_its.ps<-dads_5asv.ps

load("Data/16S/filtered_data/dadsexp.RData")
dads_16S.ps<-dads.ps

load("Data/16S/filtered_data/dads_5asv.RData")
dads_5asv_16S.ps<-dads_5asv.ps

load("Data/RPS10/filtered_data/dadsexp.RData")
dads_rps10.ps<-dads.ps

load("Data/RPS10/filtered_data/dads_5asv.RData")
dads_5asv_rps10.ps<-dads_5asv.ps
```

### Let's calculate alpha diversity metrics for ITS dataset
```{r, combine alpha diversity boxplots its}
variables <- c("Treatment", "Soil", "Incubation")
measures <- c("Observed", "InvSimpson", "Shannon")

AD_its <- phyloseq::estimate_richness(dads_its.ps, measures = c("Observed", "Shannon", "InvSimpson"))

AD_its <- cbind(AD_its, sample_data(dads_its.ps)[rownames(AD_its), c("Treatment", "Soil", "Incubation")])

variables <- c("Treatment", "Soil", "Incubation")

plots_its <- list()

for (variable in variables) {
  p <- phyloseq::plot_richness(dads_its.ps, x = variable, measures = c("Observed", "Shannon", "InvSimpson")) + 
    geom_boxplot() +
    theme_classic() +  # Use theme_classic() as a starting point
    theme(
      panel.border = element_rect(color = "black", fill = NA, size = 1),  # Add border around plot panel
      axis.line = element_blank(),  # Remove axis lines (they'll be replaced by the panel border)
      plot.margin = unit(c(1, 1, 1, 1), "lines"),  # Add some margin around the plot
      plot.title = element_text(margin = margin(b = 10)),  # Add some space below the title
      axis.title.x = element_text(margin = margin(t = 10)),  # Add some space above the x-axis title
      axis.title.y = element_text(margin = margin(r = 10))  # Add some space to the right of the y-axis title
    )
  plots_its[[variable]] <- p
}

#Let's calculate Tukey results
# Function to perform Tukey HSD and extract letters
get_tukey_letters <- function(data, measure, variable) {
  formula <- as.formula(paste(measure, "~", variable))
  aov_result <- aov(formula, data = data)
  tukey_result <- HSD.test(aov_result, variable, group = TRUE)
  letters <- tukey_result$groups$groups
  return(letters)
}

# Perform Tukey HSD tests and store results
tukey_results_ITS <- list()
for (variable in variables) {
  for (measure in measures) {
    tryCatch({
      letters <- get_tukey_letters(AD_its, measure, variable)
      tukey_results_ITS[[paste(measure, variable, sep = "_")]] <- letters
    }, error = function(e) {
      message("Error in Tukey test for ", measure, " by ", variable, ": ", e$message)
    })
  }
}


print(tukey_results_ITS)
```

### Let's calculate alpha diversity metrics for 16S dataset
```{r, combine alpha diversity boxplots 16S}

variables <- c("Treatment", "Soil", "Incubation")
measures <- c("Observed", "InvSimpson", "Shannon")

AD_16S <- phyloseq::estimate_richness(dads_16S.ps, measures = c("Observed", "Shannon", "InvSimpson"))
rownames(AD_16S) <- gsub("\\X", "", rownames(AD_16S))


# Combine richness estimates with sample metadata
AD_16S <- cbind(AD_16S, sample_data(dads_16S.ps)[rownames(AD_16S), c("Treatment", "Soil", "Incubation")])

variables <- c("Treatment", "Soil", "Incubation")

# Create an empty list to store plots
plots_16S <- list()

# Loop through each variable and create the plot
for (variable in variables) {
  p <- phyloseq::plot_richness(dads_16S.ps, x = variable, measures = c("Observed", "Shannon", "InvSimpson")) + 
    geom_boxplot() +
    theme_classic() +  # Use theme_classic() as a starting point
    theme(
      panel.border = element_rect(color = "black", fill = NA, size = 1),  # Add border around plot panel
      axis.line = element_blank(),  # Remove axis lines (they'll be replaced by the panel border)
      plot.margin = unit(c(1, 1, 1, 1), "lines"),  # Add some margin around the plot
      plot.title = element_text(margin = margin(b = 10)),  # Add some space below the title
      axis.title.x = element_text(margin = margin(t = 10)),  # Add some space above the x-axis title
      axis.title.y = element_text(margin = margin(r = 10))  # Add some space to the right of the y-axis title
    )
  plots_16S[[variable]] <- p
}

#Let's calculate Tukey results
# Function to perform Tukey HSD and extract letters
get_tukey_letters <- function(data, measure, variable) {
  formula <- as.formula(paste(measure, "~", variable))
  aov_result <- aov(formula, data = data)
  tukey_result <- HSD.test(aov_result, variable, group = TRUE)
  letters <- tukey_result$groups$groups
  return(letters)
}

# Perform Tukey HSD tests and store results
tukey_results_16S <- list()
for (variable in variables) {
  for (measure in measures) {
    tryCatch({
      letters <- get_tukey_letters(AD_16S, measure, variable)
      tukey_results_16S[[paste(measure, variable, sep = "_")]] <- letters
    }, error = function(e) {
      message("Error in Tukey test for ", measure, " by ", variable, ": ", e$message)
    })
  }
}


print(tukey_results_16S)
```

### Let's calculate alpha diversity metrics for RPS10 dataset
```{r, combine alpha diversity boxplots rps10}
variables <- c("Treatment", "Soil", "Incubation")
measures <- c("Observed", "InvSimpson", "Shannon")

# Estimate alpha diversity
AD_RPS10 <- estimate_richness(dads_rps10.ps, measures = c("Observed", "Shannon", "InvSimpson"))

# Combine richness estimates with sample metadata
AD_RPS10 <- cbind(AD_RPS10, sample_data(dads_rps10.ps)[rownames(AD_RPS10), c("Treatment", "Soil", "Incubation")])

variables <- c("Treatment", "Soil", "Incubation")

# Create an empty list to store plots
plots_rps10 <- list()

# Loop through each variable and create the plot
for (variable in variables) {
  p <- phyloseq::plot_richness(dads_rps10.ps, x = variable, measures = c("Observed", "Shannon", "InvSimpson")) + 
    geom_boxplot() +
    theme_classic() +  # Use theme_classic() as a starting point
    theme(
      panel.border = element_rect(color = "black", fill = NA, size = 1),  # Add border around plot panel
      axis.line = element_blank(),  # Remove axis lines (they'll be replaced by the panel border)
      plot.margin = unit(c(1, 1, 1, 1), "lines"),  # Add some margin around the plot
      plot.title = element_text(margin = margin(b = 10)),  # Add some space below the title
      axis.title.x = element_text(margin = margin(t = 10)),  # Add some space above the x-axis title
      axis.title.y = element_text(margin = margin(r = 10))  # Add some space to the right of the y-axis title
    )
  plots_rps10[[variable]] <- p
}

#Let's calculate Tukey results
# Function to perform Tukey HSD and extract letters
get_tukey_letters <- function(data, measure, variable) {
  formula <- as.formula(paste(measure, "~", variable))
  aov_result <- aov(formula, data = data)
  tukey_result <- HSD.test(aov_result, variable, group = TRUE)
  letters <- tukey_result$groups$groups
  return(letters)
}

# Perform Tukey HSD tests and store results
tukey_results_RPS10 <- list()
for (variable in variables) {
  for (measure in measures) {
    tryCatch({
      letters <- get_tukey_letters(AD_RPS10, measure, variable)
      tukey_results_RPS10[[paste(measure, variable, sep = "_")]] <- letters
    }, error = function(e) {
      message("Error in Tukey test for ", measure, " by ", variable, ": ", e$message)
    })
  }
}


print(tukey_results_RPS10)
```

### Let's combine the treatment plots
```{r, combine all alpha trt box plots}
create_label <- function(text, x = 0.5, y = 0.95) {  # Centered the label horizontally
  ggdraw() + 
    draw_label(text, fontface = 'bold', x = x, y = y, hjust = 0.5, vjust = 1)  # Changed hjust to 0.5 for center alignment
}

label_A <- create_label("A.")
label_B <- create_label("B.")
label_C <- create_label("C.")

# Arrange plots
plots_arrange_its <- gridExtra::grid.arrange(
  plots_its[["Treatment"]]
)
plots_arrange_16S <- gridExtra::grid.arrange(
  plots_16S[["Treatment"]]
)
plots_arrange_rps10 <- gridExtra::grid.arrange(
  plots_rps10[["Treatment"]]
)

combined_plots_16S <- plot_grid(label_A, plots_arrange_16S, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_its <- plot_grid(label_B, plots_arrange_its, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_rps10 <- plot_grid(label_C, plots_arrange_rps10, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width

facetted_figure_trt <- plot_grid(
  combined_plots_16S,
  combined_plots_its,
  combined_plots_rps10,
  nrow = 3,
  align = 'v',
  axis = 'l'
)


facetted_figure_trt_with_margins <- plot_grid(
  NULL, facetted_figure_trt, NULL,
  ncol = 3,
  rel_widths = c(0, 1, 0.08)  # 5% margin on each side
)

facetted_figure_trt_with_margins
# Save the figure
ggsave("Figures/DADS/facetted_figure_alphadiv_treatment.svg", facetted_figure_trt_with_margins, width=8, height=12)
ggsave("Figures/DADS/facetted_figure_alphadiv_treatment.pdf", facetted_figure_trt_with_margins, width=8, height=12)
```

### Let's combine the soil type plots
```{r, combine all alpha soil box plots}
create_label <- function(text, x = 0.5, y = 0.95) {  # Centered the label horizontally
  ggdraw() + 
    draw_label(text, fontface = 'bold', x = x, y = y, hjust = 0.5, vjust = 1)  # Changed hjust to 0.5 for center alignment
}

label_A <- create_label("A.")
label_B <- create_label("B.")
label_C <- create_label("C.")

# Arrange plots
plots_arrange_its <- gridExtra::grid.arrange(
  plots_its[["Soil"]]
)
plots_arrange_16S <- gridExtra::grid.arrange(
  plots_16S[["Soil"]]
)
plots_arrange_rps10 <- gridExtra::grid.arrange(
  plots_rps10[["Soil"]]
)

combined_plots_16S <- plot_grid(label_A, plots_arrange_16S, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_its <- plot_grid(label_B, plots_arrange_its, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_rps10 <- plot_grid(label_C, plots_arrange_rps10, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width

facetted_figure_soil <- plot_grid(
  combined_plots_16S,
  combined_plots_its,
  combined_plots_rps10,
  nrow = 3,
  align = 'v',
  axis = 'l'
)


facetted_figure_soil_with_margins <- plot_grid(
  NULL, facetted_figure_soil, NULL,
  ncol = 3,
  rel_widths = c(0, 1, 0.08)  # 5% margin on each side
)


facetted_figure_soil_with_margins 
# Save the figure
ggsave("Figures/DADS/facetted_figure_alphadiv_soil.svg", facetted_figure_soil_with_margins, width=8, height=12)
ggsave("Figures/DADS/facetted_figure_alphadiv_soil.pdf", facetted_figure_soil_with_margins, width=8, height=12)
```

### Let's combine the incubation plots
```{r, combine all alpha incubation box plots}
create_label <- function(text, x = 0.5, y = 0.95) {  # Centered the label horizontally
  ggdraw() + 
    draw_label(text, fontface = 'bold', x = x, y = y, hjust = 0.5, vjust = 1)  # Changed hjust to 0.5 for center alignment
}

label_A <- create_label("A.")
label_B <- create_label("B.")
label_C <- create_label("C.")

# Arrange plots
plots_arrange_its <- gridExtra::grid.arrange(
  plots_its[["Incubation"]]
)
plots_arrange_16S <- gridExtra::grid.arrange(
  plots_16S[["Incubation"]]
)
plots_arrange_rps10 <- gridExtra::grid.arrange(
  plots_rps10[["Incubation"]]
)

combined_plots_16S <- plot_grid(label_A, plots_arrange_16S, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_its <- plot_grid(label_B, plots_arrange_its, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width
combined_plots_rps10 <- plot_grid(label_C, plots_arrange_rps10, ncol = 2, rel_widths = c(0.1, 1))  # Reduced label width

facetted_figure_inc <- plot_grid(
  combined_plots_16S,
  combined_plots_its,
  combined_plots_rps10,
  nrow = 3,
  align = 'v',
  axis = 'l'
)


facetted_figure_inc_with_margins <- plot_grid(
  NULL, facetted_figure_inc, NULL,
  ncol = 3,
  rel_widths = c(0, 1, 0.08)  # 5% margin on each side
)

facetted_figure_inc_with_margins 
# Save the figure
ggsave("Figures/DADS/facetted_figure_alphadiv_incubation.svg", facetted_figure_inc_with_margins, width=8, height=12)
ggsave("Figures/DADS/facetted_figure_alphadiv_incubation.pdf", facetted_figure_inc_with_margins, width=8, height=12)
```

### Let's proceed with ordination for the ITS dataset-PCoA
```{r, ordination plots its PCOA}
strat_rel_dads_its.ps<-transform_sample_counts(dads_5asv_its.ps, function(x) x / sum(x))

ps_PCoA_its = ordinate(strat_rel_dads_its.ps, 
                   method="PCoA", 
                   distance="bray")

plot_its_trt<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_PCoA_its ,
                         color= "Treatment") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_trt<-plot_its_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_its_trt

plot_its_inc<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_PCoA_its ,
                         color= "Incubation") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_inc<-plot_its_inc +
   stat_ellipse(linetype =2) +
  theme_bw()

plot_ellipses_its_inc

plot_its_soil<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_PCoA_its ,
                         color= "Soil") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_soil<-plot_its_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_its_soil
```


### Let's proceed with ordination for the 16S dataset-PCoA
```{r, ordination plots 16s PCOA}
strat_rel_dads_16S.ps<-transform_sample_counts(dads_5asv_16S.ps, function(x) x / sum(x))


ps_PCoA_16S = ordinate(strat_rel_dads_16S.ps, 
                   method="PCoA", 
                   distance="bray")

plot_16S_trt<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_PCoA_16S ,
                         color= "Treatment") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_trt<-plot_16S_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_trt

plot_16S_inc<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_PCoA_16S ,
                         color= "Incubation") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_inc<-plot_16S_inc +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_inc

plot_16S_soil<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_PCoA_16S ,
                         color= "Soil") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_soil<-plot_16S_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_soil
```

### Let's proceed with ordination for the RPS10 dataset-pCoA
```{r, ordination plots rps10 PCOA}
strat_rel_dads_rps10.ps<-transform_sample_counts(dads_5asv_rps10.ps, function(x) x / sum(x))

ps_PCoA_rps10 = ordinate(strat_rel_dads_rps10.ps, 
                   method="PCoA", 
                   distance="bray")

plot_rps10_trt<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_PCoA_rps10 ,
                         color= "Treatment") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_trt<-plot_rps10_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_trt

plot_rps10_inc<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_PCoA_rps10 ,
                         color= "Incubation") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_inc<-plot_rps10_inc +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_inc

plot_rps10_soil<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_PCoA_rps10 ,
                         color= "Soil") +
  #ggtitle("PCoA ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_soil<-plot_rps10_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_soil
```

### Let's combine plots into facetted plot
```{r, box plots ordination pcoa}
### Let's combine plots into facetted plot
create_column <- function(ord_plot) {
  plot_grid(
    ord_plot,
    nrow = 1,
    align = 'v',
    axis = 'l',
    rel_heights = c(1)
  )
}

# Create columns for each dataset and factor
col_16S_treatment <- create_column(plot_ellipses_16S_trt)
col_16S_soil <- create_column(plot_ellipses_16S_soil)
col_16S_incubation <- create_column(plot_ellipses_16S_inc)
col_its_treatment <- create_column(plot_ellipses_its_trt)
col_its_soil <- create_column(plot_ellipses_its_soil)
col_its_incubation <- create_column(plot_ellipses_its_inc)
col_rps10_treatment <- create_column(plot_ellipses_rps10_trt)
col_rps10_soil <- create_column(plot_ellipses_rps10_soil)
col_rps10_incubation <- create_column(plot_ellipses_rps10_inc)

# Function to create headers with horizontal lines above
create_header <- function(text) {
  ggplot() +
    geom_hline(yintercept = 0.8, color = "black", size = 0.5) +
    geom_text(aes(x = 0.02, y = 0.4, label = text), size = 5, fontface = "bold", hjust = 0) +
    theme_void() +
    theme(plot.margin = margin(5, 0, 0, 10, "pt")) +
    xlim(0, 1) + ylim(0, 1)
}

# Create row plots with labels
row_16S <- plot_grid(col_16S_treatment, col_16S_soil, col_16S_incubation, 
                     ncol = 3, labels = "A.", label_size = 14, hjust = -0.5, vjust = 2)
row_its <- plot_grid(col_its_treatment, col_its_soil, col_its_incubation, 
                     ncol = 3, labels = "B.", label_size = 14, hjust = -0.5, vjust = 2)
row_rps10 <- plot_grid(col_rps10_treatment, col_rps10_soil, col_rps10_incubation, 
                       ncol = 3, labels = "C.", label_size = 14, hjust = -0.5, vjust = 2)

# Combine all elements
facetted_figure <- plot_grid(
  row_16S,
  row_its,
  row_rps10,
  nrow = 3,
  align = 'v',
  axis = 'l',
  rel_heights = c(1.1, 1, 1)  # Slightly more height for the first row
)

facetted_figure

# Save the figure
ggsave("Figures/DADS/facetted_figure_with_ordination_PCoA.svg", facetted_figure, width = 12, height = 12)
ggsave("Figures/DADS/facetted_figure_with_ordination_PCoA.pdf", facetted_figure, width = 12, height = 12)
```

### Let's proceed with ordination for the ITS dataset-NMDS
```{r, ordination plots its nmds}
strat_rel_dads_its.ps<-transform_sample_counts(dads_5asv_its.ps, function(x) x / sum(x))

ps_NMDS_its = ordinate(strat_rel_dads_its.ps, 
                   method="NMDS", 
                   distance="bray")

plot_its_trt<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_NMDS_its ,
                         color= "Treatment") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_trt<-plot_its_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_its_trt

plot_its_inc<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_NMDS_its ,
                         color= "Incubation") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_inc<-plot_its_inc +
   stat_ellipse(linetype =2) +
  theme_bw()

plot_ellipses_its_inc

plot_its_soil<- plot_ordination(strat_rel_dads_its.ps, 
                         ps_NMDS_its ,
                         color= "Soil") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_its_soil<-plot_its_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_its_soil
```

### Let's proceed with ordination for the 16S dataset-NMDS
```{r, ordination plots 16s nmds}
strat_rel_dads_16S.ps<-transform_sample_counts(dads_5asv_16S.ps, function(x) x / sum(x))


ps_NMDS_16S = ordinate(strat_rel_dads_16S.ps, 
                   method="NMDS", 
                   distance="bray")

plot_16S_trt<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_NMDS_16S ,
                         color= "Treatment") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_trt<-plot_16S_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_trt

plot_16S_inc<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_NMDS_16S ,
                         color= "Incubation") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_inc<-plot_16S_inc +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_inc

plot_16S_soil<- plot_ordination(strat_rel_dads_16S.ps, 
                         ps_NMDS_16S ,
                         color= "Soil") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_16S_soil<-plot_16S_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_16S_soil
```

### Let's proceed with ordination for the RPS10 dataset-NMDS
```{r, ordination plots rps10 nmds}
strat_rel_dads_rps10.ps<-transform_sample_counts(dads_5asv_rps10.ps, function(x) x / sum(x))

ps_NMDS_rps10 = ordinate(strat_rel_dads_rps10.ps, 
                   method="NMDS", 
                   distance="bray")

plot_rps10_trt<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_NMDS_rps10 ,
                         color= "Treatment") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Treatment") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_trt<-plot_rps10_trt +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_trt

plot_rps10_inc<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_NMDS_rps10 ,
                         color= "Incubation") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Incubation") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_inc<-plot_rps10_inc +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_inc

plot_rps10_soil<- plot_ordination(strat_rel_dads_rps10.ps, 
                         ps_NMDS_rps10 ,
                         color= "Soil") +
  #ggtitle("NMDS ordination plot - relative abundance") +
  labs(color = "Soil type") + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=14), title =  element_text(size=13), legend.text = element_text(size = 18))

plot_ellipses_rps10_soil<-plot_rps10_soil +
   stat_ellipse(linetype = 2) +
  theme_bw()

plot_ellipses_rps10_soil
```

### Let's combine plots into facetted plot
```{r, box plots ordination nmds}
create_column <- function(ord_plot) {
  plot_grid(
    ord_plot,
    nrow = 1,
    align = 'v',
    axis = 'l',
    rel_heights = c(1)
  )
}

# Create columns for each dataset and factor
col_16S_treatment <- create_column(plot_ellipses_16S_trt)
col_16S_soil <- create_column(plot_ellipses_16S_soil)
col_16S_incubation <- create_column(plot_ellipses_16S_inc)
col_its_treatment <- create_column(plot_ellipses_its_trt)
col_its_soil <- create_column(plot_ellipses_its_soil)
col_its_incubation <- create_column(plot_ellipses_its_inc)
col_rps10_treatment <- create_column(plot_ellipses_rps10_trt)
col_rps10_soil <- create_column(plot_ellipses_rps10_soil)
col_rps10_incubation <- create_column(plot_ellipses_rps10_inc)

# Create row plots with labels
row_16S <- plot_grid(col_16S_treatment, col_16S_soil, col_16S_incubation, 
                     ncol = 3, labels = "A.", label_size = 14, hjust = -0.5, vjust = 2)
row_its <- plot_grid(col_its_treatment, col_its_soil, col_its_incubation, 
                     ncol = 3, labels = "B.", label_size = 14, hjust = -0.5, vjust = 2)
row_rps10 <- plot_grid(col_rps10_treatment, col_rps10_soil, col_rps10_incubation, 
                       ncol = 3, labels = "C.", label_size = 14, hjust = -0.5, vjust = 2)

# Combine all elements
facetted_figure <- plot_grid(
  row_16S,
  row_its,
  row_rps10,
  nrow = 3,
  align = 'v',
  axis = 'l',
  rel_heights = c(1.1, 1, 1)  # Slightly more height for the first row
)

facetted_figure

# Save the figure
ggsave("Figures/DADS/facetted_figure_with_ordination_NMDS.svg", facetted_figure, width = 12, height = 12)
ggsave("Figures/DADS/facetted_figure_with_ordination_NMDS.pdf", facetted_figure, width = 12, height = 12)
```
